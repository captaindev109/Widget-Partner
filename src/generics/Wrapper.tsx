import { FC, ReactNode } from 'react';
import styled from 'styled-components';
import Card from './Card';

export interface WrapperProps {
  children: ReactNode;
  direction?: 'row' | 'column';
  display?: string;
  className?: string;
}

const StyledWrapper = styled(Card)<WrapperProps>`
  display: ${({ display }) => display || 'block'};
  flex-direction: ${({ direction }) => direction || 'row'};
`;

const Wrapper: FC<WrapperProps> = (props) => {
  const { children, className, ...newProps } = props;
  return (
    <StyledWrapper className={className} {...newProps} data-testid="wrapper">
      {children}
    </StyledWrapper>
  );
};

Wrapper.defaultProps = { direction: 'row' };

export default Wrapper;
