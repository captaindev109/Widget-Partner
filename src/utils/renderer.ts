import React from 'react';
import { ConfigProps, Index, EachComponentObjectChildren, BasicAny } from '@/types/types';
import { UIComponents } from './components';

const recursivelyCreateReactComponents = (
  componentToCreate: BasicAny,
  configuration: ConfigProps,
  children: BasicAny
) => {
  if (typeof componentToCreate !== 'undefined') {
    return React.createElement(
      componentToCreate,
      configuration,
      children &&
        (typeof children === 'string'
          ? children
          : children.map((component: EachComponentObjectChildren, index: Index) => viewRenderer(component, index)))
    );
  }
  null;
};

const viewRenderer: BasicAny = (config: ConfigProps, index: Index) => {
  const children = config.children;
  const newConfig = { ...config, key: index };
  const LookForComponentInUILibrary = UIComponents[config.component];
  return recursivelyCreateReactComponents(LookForComponentInUILibrary, newConfig, children);
};

export default viewRenderer;
