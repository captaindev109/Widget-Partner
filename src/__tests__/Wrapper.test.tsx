import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

import { FC } from 'react';
import Wrapper from '@/generics/Wrapper';

const child_text = 'child';

const WrapperComponent: FC = () => <Wrapper>{child_text}</Wrapper>;

describe('Render tests', () => {
  test('to see if the render its children', () => {
    const { getByTestId } = render(<WrapperComponent />);
    const wrapper = getByTestId('wrapper');
    expect(wrapper).toHaveTextContent(child_text);
  });

  it('should match the sapshot', () => {
    const snap = renderer.create(<WrapperComponent />).toJSON();
    expect(snap).toMatchSnapshot();
  });
});

describe('Prop tests', () => {
  test('to see if default props is row', () => {
    const { getByTestId } = render(<WrapperComponent />);
    const wrapper = getByTestId('wrapper');
    expect(wrapper).toHaveStyle(`flex-direction: row`);
  });

  it('to see if direction works is row', () => {
    const { getByTestId } = render(<Wrapper direction="column">{child_text}</Wrapper>);
    const wrapper = getByTestId('wrapper');
    expect(wrapper).toHaveStyle(`flex-direction: column`);
  });
});
