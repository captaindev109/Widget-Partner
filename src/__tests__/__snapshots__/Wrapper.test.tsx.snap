// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render tests should match the sapshot 1`] = `
<div
  className="Card-sc-fb64gk-0 Wrapper__StyledWrapper-sc-drqeso-0 leCJCh bWoYwk"
  data-testid="wrapper"
  direction="row"
>
  child
</div>
`;
